Sources/web3swift/Transaction/AbstractEnvelope.swift

```diff
@@ -0,0 +1,157 @@
+    /// will return an new EthereumTransaction object on success
+    /// returns nil if a required field is not found in the decoder stream, or can't be decoded
+    init?(rawValue: Data) // Decode from Ethereum Data
+
+    // pseudo memberwise initializer
+    // accepts all common parameters (full description with default implementation below)
+    init(to: EthereumAddress, nonce: BigUInt?, chainID: BigUInt?, value: BigUInt?,
+         data: Data, v: BigUInt, r: BigUInt, s: BigUInt, options: TransactionOptions?)
+
+    /// Applies the passed options to the transaction envelope
+    ///   - Parameters:
+    ///     - {default}: TransactionOptions struct
+    mutating func applyOptions(_ options: TransactionOptions)
+    ///  Create a TransactionOptions object representing this transaction
+    /// - Returns: A TransactionOptions object with all parameters from the transaction set
+    func getOptions() -> TransactionOptions
+
+    /// Transaction encoder for transmission or signing
+    ///  - Parameters:
+    ///    - {default}: EncodeType enum
+    ///     - when type is .signature the transaction is encoded for signing
+    ///     - when type is .transaction the thransaction is encoded for hashing or transmission to the blockchain
+    /// - Returns: a raw encoding stream representing the transaction, encoded according to it's type
+    func encode(for type: EncodeType) -> Data?
+    /// Encodes the transaction as a set of strings for JSON transmission
+    /// - Returns: A TransactionParameters object containg all the parameters for the transaction
+    func encodeAsDictionary(from: EthereumAddress?) -> TransactionParameters?
+
+    /// used by the signing algorithm to set the v, r, s parameters
+    /// - Parameters:
+    ///   - {default}: UnmarshalledSignature struct containing the v, r, s parameters
+    mutating func setUnmarshalledSignatureData(_ unmarshalledSignature: SECP256K1.UnmarshaledSignature)
+    /// used by the public-key recovery algorithm for determining the signer of the transaction
+    /// - Returns: returns the 'unmarshalled' v,r,s psignatrure parameters
+    func getUnmarshalledSignatureData() -> SECP256K1.UnmarshaledSignature?
+}
+
+public extension AbstractEnvelope {
+    // default implementation, should be overridden by each Envelope implementation
+    /// Default memberwse initializer that all envelopes must support
+    /// - Parameters:
+    ///   - to: EthereumAddress of destination
+    ///   - nonce: nonce for the transaction
+    ///   - chainID: chainId of the network the transaction belongs to
+    ///   - value: Native value in Wei of the transaction
+    ///   - data: Payload data for the transaction
+    ///   - v: Signature V component
+    ///   - r: Signature R component
+    ///   - s: Signature S component
+    ///   - options: TransactionOptions struct containing any other required parameters
+    init(to: EthereumAddress, nonce: BigUInt? = nil,
```

Comment:
yes this builds, suitable defaults are provided in the actual implementation. Allowing for nil here, allows for the value to be pulled from the passed options, if present. There's a hierarchy of resolution direct > options > default

Reply:
As i said i've found the way how you make this works. And as i said it's ok for me the chain that you're making, but i disagree the tool that you're taking).

I see that default implementation in protocol should be used as exact default implementation in final struct. E.g. code in protocol extension should be enough to use this initializer as is and not to re implement them there. 

I suggest you to drop this implementation within extension, since it's not used anywhere as i see (you've re implementing it in all three final struct). The better way as for me is to make default generic init which takes in `options` in account, but it'll brings demand to make some additional protocols with the requirement to the `TransactionOptions`, so i think it worth single PR for that.

Reply:
True that the version here is redundant, as all it does is chain forward to the final implementation. I will remove this one. and then move the doc-comment up to where it's declared in the protocol

Reply:
Done, will appear in next commit